src/BinSkim.Rules/PERules/BA6001.DisableIncrementalLinking.cs

```diff
@@ -0,0 +1,93 @@
+using Microsoft.CodeAnalysis.BinaryParsers.ProgramDatabase;
+using Microsoft.CodeAnalysis.IL.Sdk;
+using Microsoft.CodeAnalysis.Sarif;
+using Microsoft.CodeAnalysis.Sarif.Driver;
+
+namespace Microsoft.CodeAnalysis.IL.Rules
+{
+    [Export(typeof(ReportingDescriptor))]
+    [Export(typeof(Skimmer<BinaryAnalyzerContext>))]
+    public class DisableIncrementalLinking : WindowsBinaryAndPdbSkimmerBase
+    {
+        /// <summary>
+        /// BA6001
+        /// </summary>
+        public override string Id => RuleIds.DisableIncrementalLinking;
+
+        /// <summary>
+        /// Incremental linking support increases binary size and can reduce runtime performance. Fully optimized 
+        /// release builds should not specify incremental linking.
+        /// </summary>
+
+        public override MultiformatMessageString FullDescription =>
+            new MultiformatMessageString { Text = RuleResources.BA6001_DisableIncrementalLinking_Description };
+
+        protected override IEnumerable<string> MessageResourceNames => new string[] {
+                    nameof(RuleResources.BA6001_Pass),
+                    nameof(RuleResources.BA6001_Warning),
+                    nameof(RuleResources.BA6001_Pass_NonReleaseBuild)
+                };
+
+        private const string AnalyzerName = RuleIds.DisableIncrementalLinking + "." + nameof(DisableIncrementalLinking);
+
+        public override AnalysisApplicability CanAnalyzePE(PEBinary target, Sarif.PropertiesDictionary policy, out string reasonForNotAnalyzing)
+        {
+            PE portableExecutable = target.PE;
+            AnalysisApplicability result = AnalysisApplicability.NotApplicableToSpecifiedTarget;
+
+            reasonForNotAnalyzing = MetadataConditions.ImageIsILOnlyAssembly;
+            if (portableExecutable.IsILOnly) { return result; }
+
+            reasonForNotAnalyzing = MetadataConditions.ImageIsResourceOnlyBinary;
+            if (portableExecutable.IsResourceOnly) { return result; }
+
+            reasonForNotAnalyzing = MetadataConditions.CouldNotLoadPdb;
+            if (target.Pdb == null) { return result; }
+
+            reasonForNotAnalyzing = null;
+            return AnalysisApplicability.ApplicableToSpecifiedTarget;
+        }
+
+        public override void AnalyzePortableExecutableAndPdb(BinaryAnalyzerContext context)
+        {
+            PEBinary target = context.PEBinary();
+            PE pe = target.PE;
+            Pdb pdb = target.Pdb;
+
+            if (!pe.IsMostlyOptimized(pdb))
+            {
+                context.Logger.Log(this,
+                    RuleUtilities.BuildResult(ResultKind.NotApplicable, context, null,
+                    nameof(RuleResources.BA6001_Pass_NonReleaseBuild),
+                    context.TargetUri.GetFileName()));
+                return;
+            }
+
+            if (pe.IncrementalLinkingEnabled(pdb))
+            {
+                context.Logger.Log(this,
```

Comment:
>Logger [](http://example.com/codeflow?start=24&length=6)

Generally, when we log, we provide a code comment that has the complete text of the message that is output to the users, basically, recap the localized message string.

This helps when reading the code and maximizes chances devs will go fix up the messages to improve or update them when logic is changed. The downside is message drift, we need to make sure both copies stay in sync.

Go take a look at any existing security rule, I think we've mostly got them all in shape for this code convention.

Reply:
Will do.